### ISSUE TITLE
compiler support 

### ISSUE THREAD

**Issue by shahzebsiddiqui** [HPC-buildtest/schemas#25]
It would be nice to have buidtest support compiler types which picks the appropriate compiler wrapper and just alters variable ``cc``, ``fc``, ``cxx``, ``cflags``, ``cxxflags``, ``ldflags``. For initial support we can try to support ``gnu``, ``intel``, ``pgi``, ``cuda`` and ``clang``.  

The way compilers could work is take a single test and allow user to compile with different compilers so that test doesn't have to be duplicated.

Here is a typical example for compiling vector addition with different compilers

```
version: 0.0.1
hello_gnu:
  type: script
  description: vector addition using gnu compiler
  env: 
    cc: gcc
    cflags: -O2
    ldflags: -lm
    srcdir: apps
    exe: vecAdd
  run: |
    module load gcc
    $cc $cflags -o $exe $srcdir/vecAdd.c $ldflags
    ./$exe 
 
hello_intel:
  type: script
  description: vector addition using intel compiler
  env: 
    cc: icc
    cflags: -O2
    ldflags: -lm
    srcdir: apps
    exe: vecAdd
  run: |
    module load intel
    $cc $cflags -o $exe $srcdir/vecAdd.c $ldflags
    ./$exe 
```

Given that above code is repetitive, we should allow some form of compiler selection that sets these variable such that they can be referenced in ``run`` section. I think buildtest should use the ``compiler`` key to just set the variables similar to what is done in ``env`` section that declares environment variable. 

Here is a proposal schema that supports gnu and intel compiler. The required key is ``cc``, ``fc`` or ``cxx`` so user can specify which one they want based on source file they compiling. We have a ``module`` key to allow user to specify a module to load based on the compiler. Down side with this approach is we can't specify multiple gnu compilers to test.

```
version: 0.0.1
hello_compiler_example:
  type: script
  description: vector addition using gnu and intel compiler
  compiler:
    gnu:
      module: |
          module purge
          module load gcc/7.0
      cc: gcc
      cflags: -O2
      ldflags: -lm
    intel:
      module: |
        module purge
        module load intel/18     
      cc: icc
      cflags: -O2
      ldflags: -lm
  env: 
    srcdir: apps
    exe: vecAdd
  run: |
    $cc $cflags -o $exe $srcdir/vecAdd.c $ldflags
    ./$exe 
```

Here is an approach for a more nested compiler structure. Inside ``gnu`` and ``intel`` we have a list of objects where each object is a pattern type such as ``gcc/7.0`` and ``gcc/6.0``. These names would be tacked on to the test or part of the output report.

With this approach we can support multiple compiler instances but top level key like ``gnu``, ``intel`` would have to be pre-defined.

```
version: 0.0.1
hello_compiler_example:
  type: script
  description: vector addition using gnu and intel compiler
  compiler:
    gnu:
    - gcc/7.0: 
        module: |
            module purge
            module load gcc/7.0
        cc: gcc
        cflags: -O2
        ldflags: -lm

   - gcc/6.0:
       module: |
            module purge
            module load gcc/6.0
       cc: gcc
       cflags: -O1 -Wall
       ldflags: -lm

    intel:
      intel/18: 
        module: |
          module purge
          module load intel/18     
        cc: icc
        cflags: -O2
        ldflags: -lm
  env: 
    srcdir: apps
    exe: vecAdd
  run: |
    $cc $cflags -o $exe $srcdir/vecAdd.c $ldflags
    ./$exe 
```

Here is an extension to previous design but with ability to avoid redefining variables if they are same across compiler group (``gnu``) or all compilers via use of ``commons`` key. 

```
version: 0.0.1
hello_compiler_example:
  type: script
  description: vector addition using gnu and intel compiler
  compiler:
    gnu:
      commons:
        cc: gcc
        cflags: -O2
        ldflags: -lm
      spec:
      - gcc/7.0: 
           module: |
             module purge
             module load gcc/7.0
     - gcc/6.0:
          module: |
            module purge
            module load gcc/6.0       
       cflags: -O1 -Wall

    intel:
      spec:
      - intel/18: 
          module: |
            module purge
            module load intel/18     
          cc: icc
          cflags: -O2
          ldflags: -lm
  env: 
    srcdir: apps
    exe: vecAdd
  run: |
    $cc $cflags -o $exe $srcdir/vecAdd.c $ldflags
    ./$exe 
```

The ``commons`` is a shared across group of gnu or intel it's optional, but user can override commons in the case of ``gcc/6.0`` where ``cflags`` was set to ``-O1 -Wall``. 

In this case we had to introduce a new key in this case ``spec`` which is under each compiler group which is a  list of objects.

I realize this can get more complex with what it already is, but this is an idea what i had in mind. The ``module`` key is used to specify a list of modules to load as is. This would get injected into the ``run`` section but at the moment user has no control in what order this is placed. Generally modules are placed right after ``#SBATCH`` directives but user is not aware of this.

**Comment by vsoch** [comment]
I don't like any of the more complex examples - they are very complicated (in terms of a config) and looking back over them, your first base cases (that don't require any new support) are the most clear, straight forward, and (imho) easiest to implement. If we implement some definition for compilers, I can assure you it will be error prone, and abstract away from what the user really wants to see (and control) the command. Also, if you are making an argument for brevity, the later examples you give are actually a lot longer and more complicated.  Did you notice that? :) The first two are:

 - the shortest ones
 - easiest to control and customize
 - much less abstraction and likely to run into issues

So TLDR I think (for now) adding compiler support is akin to over-engineering, and to "add compiler support" we should simply add a really good example recipe for each intended type.

**Comment by shahzebsiddiqui** [comment]
Yea i know shorter the better. I have some test examples in https://github.com/HPC-buildtest/buildtest-stampede2 for compiler and will be adding some more.

I am not sure if adding a new ``type`` would be warranted for this. This is kind of related to the module permutation we discussed in https://github.com/HPC-buildtest/buildtest-framework/issues/275. The only difference here is we have a single test that tested for multiple compiler group or even multiple compiler versions. 

I have struggled to find an elegant way to write a YAML for supporting multiple compilers other than just duplicating the test which is not efficient. How would you approach this problem.

**Comment by vsoch** [comment]
I don't think a new type is appropriate either, they are just scripts.

I think, for the purposes of reproducibility, if you can't accomplish what you want with a build matrix, then duplication is okay. I choose simple and duplicated over complex, error prone, and not.

**Comment by shahzebsiddiqui** [comment]
@vsoch i'd like to add the ``compiler`` key and i can see this to be an extension to the ``type: script`` for now. My only concern i have is how to use variables in YAML to reference something in ``run`` section. I can see Node Anchors in YAML as the closest thing to a variable but YAML doesn't support variable substitution. 

What i'd like to figure out is how to reference **module**, and **source**  to reference a file. Should it be ``{{compiler.gnu.module}}`` to reference module section in compiler gnu or something else. I can't figure out how i can reference data in run section when shared between multiple compilers.

```
version: 0.0.1
hello_compiler_example:
  type: script
  description: vector addition using gnu and intel compiler
  compiler:
    source: src/vecAdd.c
    gnu:
      module: |
          module purge
          module load gcc/7.0
      cflags: -O2
      ldflags: -lm
    intel:
      module: |
        module purge
        module load intel/18     
      cflags: -O2
      ldflags: -lm
  env: 
    exe: vecAdd
  run: |
    {{compiler.gnu.module}}
    $cc   {{compiler.gnu.cflags}} -o $exe ${{compiler.source}}     {{compiler.gnu.ldflags}}
    ./$exe 
```

I think buildtest can easily figure out the compiler wrapper based on file extension of the source file so user just needs to reference ``$cc`` ``$fc`` or ``$cxx`` to get the c, c++ or fortran compiler. 

Any thoughts how you would design this YAML example.

**Comment by vsoch** [comment]
I don't think compiler matches logically with script - a script implies a text file... script. A compiler implies - something compiled to a binary. I think both cases should be supported, but as separate types.  If you want to start with script as a base for a new type that is most logical, and go from there.

**Comment by shahzebsiddiqui** [comment]
hmm that's what I was my other thought so a new type let's say ``type: compiler`` basically to address compiler builds. I think it would need ``run`` as a type. What about this idea

```
version: 0.0.1
hello_world:
  type: compiler
  module:
    gnu: |
       module purge
       module load gcc
    intel: |
       module purge
       module load intel/17
  compiler: 
    source: src/hello.c
    gnu:
      cflags: -O1
    intel:
     cflags: -O1   
``` 

This is a bit simple example, i guess scope of the ``type: compiler`` is subject to compiling single source files and nothing to fancy. The ``source`` key is to specify the source file relative to Buildspec file. I'd need some way to specify the modules in order to load them when writing the script.

**Comment by vsoch** [comment]
I like how that looks - very clean! Can you explain to me the distrinction between having module load commands under gnu vs. intel? Is the idea that the executor would handle something under the hood?

**Comment by shahzebsiddiqui** [comment]
I anticipate the above script will generate two tests as follows

```
module purge 
module load gcc
gcc -O1 <path>/src/hello.c
./a.out
```

```
module purge 
module load intel/17
icc -O1 <path>/src/hello.c
./a.out
```

I think the executor should manage how the testscript is to be run but not how the test is created. What i needed to figure out is a way for users to specify module load based on the compiler selection. If modules are not defined in Buildspec then we assume ``module load`` commands are not inserted and we assume user has the compiler in $PATH set somewhere. For now had ``module`` as an object which has a compiler name as identifier to specify what modules to load.

We could even simplify the module to just accept the name of module as pose to writing the actual module command

```
module:
  gnu:
    purge: yes
    name: ["gcc"]
  intel:
    purge: no
    name: ["intel/17"]
```

This above example means with ``purge: yes`` to run ``module purge`` and ``name`` is a list of modules to load. 

There is a difference between the first and second example. First example it does not guarantee all modules X Y Z are loaded but second example guarantees all modules X Y Z are loaded.

```
module load X Y Z
```

```
module load X
module load Y 
module load Z

# this is equivalent to
module load X && module load Y && module load Z
```

**Comment by vsoch** [comment]
If the modules are loaded the same regardless of the final compile command, why do they need to have named sections based on the compiler? Wouldn't a simple "module" section (that can be used across compilers) be much simpler? The user should know that yes, given something compiling with gcc, they should add the needed modules to  load for that. Same for intel, etc.

**Comment by shahzebsiddiqui** [comment]
I am in all favors of a more simple design and ``module`` is not a required key but i foresee the module section would be used since i'd want to test a single test with a specific compiler version.

Another idea would be as follows

```
module:
  gnu: "module purge && module load gcc"
  intel: "module purge && module load intel/17" 
```

Imagine a cluster with many compiler versions and i'd want to test a single code with different versions.  To get different version we need to alter module load command.

**Comment by shahzebsiddiqui** [comment]
What about this example which creates 4 tests with 2 versions of gcc and intel. I think buildtest would need to check if ``gnu`` is defined in compiler before checking the ``module`` section.  Changing the ``gnu`` and ``intel`` key in module section from string to  list allows me to specify multiple versions. 

With this approach we assume the compilation pattern is same across all gnu and intel version.


```
version: 0.0.1
hello_world:
  type: compiler
  module:
    gnu:  
     - "module purge && module load gcc/4.0"
     - "module purge && module load gcc/6.0"
    intel: 
   -  "module purge &&  module load intel/17"
   -  "module purge &&  module load intel/18"
  compiler: 
    source: src/hello.c
    gnu:
      cflags: -O1
    intel:
     cflags: -O1   
```

**Comment by vsoch** [comment]
Oh I see, so you want to run two different tests (each with different compilers) that have different module dependencies. The simplest thing to do here is to not try and cram them into one test, but do:
```yaml
version: 0.0.1
hello_world_gnu:
  type: compiler
  module:
     - "module purge && module load gcc/4.0"
     - "module purge && module load gcc/6.0"
  compiler: 
    source: src/hello.c
    gnu:
      cflags: -O1
```
and then

```yaml
hello_world_intel:
  type: compiler
  module:
   -  "module purge &&  module load intel/17"
   -  "module purge &&  module load intel/18"
  compiler: 
    source: src/hello.c
    intel:
     cflags: -O1 
```
Yes, it's slightly more code, but I find it much more clear to read, and if I need to debug one of the compilers, they aren't wrapped into what comes down to one test - each has it's own output file, etc. This also means the module key can be used generally across tests.

**Comment by vsoch** [comment]
And the implication is that each module line corresponds to a separate test run for the same compile/build, but under the same testing namespace of "hello_world_intel"

**Comment by shahzebsiddiqui** [comment]
Let me give this a shot and design the schema, i like this approach what you proposed stay tuned.
